ir/src/error.rs

```diff
@@ -75,4 +76,30 @@ impl SemanticError {
             matrix_col_len
         ))
     }
+
+    pub(super) fn invalid_list_folding(
+        lf_value_type: &air_script_core::ListFoldingValueType,
+        symbol_type: &IdentifierType,
+    ) -> SemanticError {
+        SemanticError::InvalidListFolding(format!(
+            "Symbol type {} is not supported for list folding value type {:?}",
+            symbol_type, lf_value_type
+        ))
+    }
+
+    pub(super) fn list_folding_empty_list(
+        lf_value_type: &air_script_core::ListFoldingValueType,
+    ) -> SemanticError {
+        SemanticError::InvalidListFolding(format!(
+            "List folding value cannot be an empty list. {:?} represents an empty list.",
+            lf_value_type
+        ))
+    }
+
+    pub(super) fn single_trace_column_allowed(ident: &str) -> SemanticError {
+        SemanticError::InvalidUsage(format!(
+            "{} should be a single trace column and not a group.",
+            ident
+        ))
+    }
```

Comment:
nit: I think there might be a better name for this. It's a type mismatch issue, so "single_trace_column_allowed" seems like a misnomer to me. Maybe "invalid_trace_binding" and the error message could be something like "Expected {} to be a binding to a single trace column."? In general, I want to move to the "binding" terminology in the future anyway

It might not be worth changing this now though. Maybe we should just revisit the errors again after this is merged and the refactors are done

Reply:
Thanks. I've changed this for now but agreed we should revisit all the errors again.